### PR TITLE
Add Push and In app notifications

### PR DESCRIPTION
**Pull request by JawadHossain**
Screenshots:
notification 
Note: delete account and recreate to make userId and documentId same in Firestore

### CODE REVIEW

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java [hunk 1]
```diff
@@ -160,9 +167,25 @@ private void btnRequestBook(View view)
         this.bookDocument.update(getString(R.string.requesters),
                 FieldValue.arrayUnion(UserCredentialAPI.getInstance().getUsername()));
         this.bookDocument.update(getString(R.string.status), getString(R.string.requested));
+
+        // Send In-app and Push notification to owner
+        sendBorrowRequestNotification();
         onBackClick(view);
     }
 
+    private void sendBorrowRequestNotification()
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java:176]: missing Javadoc header

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java [hunk 2]
```diff
@@ -160,9 +167,25 @@ private void btnRequestBook(View view)
...
+                + UserCredentialAPI.getInstance().getUsername();
+        HashMap<String, String> inAppNotification = new HashMap<>();
+        inAppNotification.put("bookID", selectedBookId);
+        inAppNotification.put("message", message);
+        inAppNotification.put("timestamp", String.valueOf(Timestamp.now().getSeconds()));
+        /* Call notification handler to process notification */
+        NotificationHandler.sendNotification("Borrow Request", message, selectedBook.getOwner(), inAppNotification);
+    }
+
     private void btnSetLocation(View view)
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java:189]: Would you mind adding a Javadoc header to this too?

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java [hunk 3]
```diff
@@ -179,6 +185,61 @@ public void onComplete(@NonNull Task<Void> task)
                 });
     }
 
+    /**
+     * Increment User notification count
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java:189]: Javadoc missing parameter

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java [hunk 4]
```diff
@@ -179,6 +185,61 @@ public void onComplete(@NonNull Task<Void> task)
...
+     */
+    public static void incrementNotificationCount(String userId)
+    {
+        FirebaseFirestore.getInstance().collection("Users")
+                .document(userId)
+                .update("notificationCount", FieldValue.increment(1));
+    }
+
+    /**
+     * Reset User notification Count and Delete Notifications
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java:199]: Javadoc missing parameters

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java [hunk 5]
```diff
@@ -179,6 +185,61 @@ public void onComplete(@NonNull Task<Void> task)
...
+     */
+    public static void resetNotifications(String userId)
+    {
+        /* Update UserCredentialAPI*/
+        UserCredentialAPI.getInstance().setNotificationCount(null);
+        /* set notification count to null */
+        firebase.collection("Users")
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java:206]: Please try and avoid hardcoding the collection names

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/MyBooksFragment.java [hunk 6]
```diff
@@ -69,6 +71,21 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
         myBooksView.findViewById(R.id.app_header_filter_button).setVisibility(View.VISIBLE);
         myBooksView.findViewById(R.id.app_header_scan_button).setVisibility(View.VISIBLE);
 
+        /* Set visibility of notification icon badge if notifications present*/
+        notificationIconBadge = myBooksView.findViewById(R.id.notification_icon_badge);
+        Long notificationCount = UserCredentialAPI.getInstance().getNotificationCount();
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/MyBooksFragment.java:76]: This looks amazing, bravo

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import retrofit2.Call;
+import retrofit2.http.Body;
+import retrofit2.http.Headers;
+import retrofit2.http.POST;
+
+public interface APIService
+{
+    @Headers(
+            {
+                    "Authorization:key=AAAAxQez4bg:APA91bHXkCnrOfTnVcmkfxC-9hqMrvvNJdHb_qEcG4-j0YxbUaHFq9kB6pzHkU4XJSJhyTt7lpXgCyXyO8jJRokJIKErI9JZVGkCRV1nIBUb1f1intxwHai1RCTrhObXX6nQ8rK4AHDN",
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java:13]: Should this key be kept in secrets.xml?

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+package com.example.bookwormadventuresdeluxe2.NotificationUtility;
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java:1]: Make sure to leave a comment describing what this class is for

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Client.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package com.example.bookwormadventuresdeluxe2.NotificationUtility;
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Client.java:1]: Missing header comment again

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Data.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package com.example.bookwormadventuresdeluxe2.NotificationUtility;
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Data.java:1]: Missing header comment again

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/NotificationData.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package com.example.bookwormadventuresdeluxe2.NotificationUtility;
+
+/**
+ * Body of the notification
+ */
+public class NotificationData
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/NotificationData.java:6]: Is this class actually doing anything?

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/MyResponse.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package com.example.bookwormadventuresdeluxe2.NotificationUtility;
+
+/**
+ * Stores request result
+ */
+public class MyResponse
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/MyResponse.java:6]: Is this class necessary?

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/RequestDetailViewFragment.java [hunk 13]
```diff
@@ -206,9 +213,30 @@ private void btnAccept(View view)
...
         this.bookDocument.update(getString(R.string.requesters), borrower);
         this.bookDocument.update(getString(R.string.status), getString(R.string.accepted));
+
+        String borrowerUsername = borrower.get(0);
+        // Send In-app and Push notification to Borrower
+        sendRequestAcceptedNotification(borrowerUsername);
         onBackClick(view);
     }
 
+    private void sendRequestAcceptedNotification(String borrowerUsername)
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/RequestDetailViewFragment.java:223]: Missing javadoc header

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/RequestDetailViewFragment.java [hunk 14]
```diff
@@ -88,10 +92,13 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
+                if (this.selectedBook.getPickUpAddress().equals(""))
+                {
                     this.btn2.setBackgroundTintList(getResources().getColorStateList(R.color.tempPhotoBackground));
                     this.btn2.setTextColor(getResources().getColorStateList(R.color.colorPrimary));
-                } else {
+                }
+                else
+                {
                     this.btn2.setOnClickListener(this::btnLendBook);
 //                    this.bookDetailView.findViewById(R.id.borrow_exchange).setVisibility(View.VISIBLE);
```
Comment (KateVlaar) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/RequestDetailViewFragment.java:103]: Can we get rid of this line?

**Review by KateVlaar — CHANGES_REQUESTED** [review]

**Comment by KateVlaar** [comment]
Can we make the text black instead of red for the notification? Also, what happens if you click on a book in the notifications tab? Are you taking to the requests page?

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java (hunk 1 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java:179]: done

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java (hunk 2 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/BorrowDetailViewFragment.java:197]: done

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java (hunk 3 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java:189]: done

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java (hunk 4 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java:201]: done

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java (hunk 5 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java]: My apologies. Was working on this at 4 AM lol. Fixed

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/RequestDetailViewFragment.java (hunk 13 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/RequestDetailViewFragment.java:243]: added.

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java (hunk 8 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java]: sorry for the delay. Renamed and added comments to all files.

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Client.java (hunk 9 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Client.java]: added.

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Data.java (hunk 10 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/Data.java:1]: added.

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/NotificationData.java (hunk 11 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/NotificationData.java]: I renamed this to NotificationSender. Pls see processNotification inside NotificationHandler for usage. The added comments should clarify.
processNotification Inside NotificationHandler

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/MyResponse.java (hunk 12 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/MyResponse.java]: I renamed this to RetrofitResponse. Pls see processNotification inside NotificationHandler for usage. The added comments should clarify.
processNotification Inside NotificationHandler

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java [hunk 15]
```diff
@@ -56,13 +62,13 @@ public void onComplete(@NonNull Task<QuerySnapshot> task)
...
                                 UserProfileObject userObject = new UserProfileObject(
-                                                    document.getData().get(context.getString(R.string.firestore_username)).toString(),
-                                                    document.getData().get(context.getString(R.string.firestore_email)).toString(),
-                                                    document.getData().get(context.getString(R.string.firestore_phoneNumber)).toString(),
-                                                    document.getData().get(context.getString(R.string.firestore_userId)).toString(),
-                                                    document.getId()
-                                                    );
-
+                                        document.getData().get(context.getString(R.string.firestore_username)).toString(),
+                                        document.getData().get(context.getString(R.string.firestore_email)).toString(),
```
Comment (JawadHossain) [inline app/src/main/java/com/example/bookwormadventuresdeluxe2/FirebaseUserGetSet.java:66]: wasn't intended.

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/MyResponse.java (hunk 12 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/MyResponse.java]: Changed notification badge text colour to black.
Currently notification clicks take user to login page or my Books if logged in.

File: app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java (hunk 7 above)
Comment (JawadHossain) [reply app/src/main/java/com/example/bookwormadventuresdeluxe2/NotificationUtility/APIService.java]: The attributes here take constants only, wasn't able to use context.getString() from secrets.xml. 
Added CloudMessagingServerKeyConstant.java to .gitignore. Pls see image:
cloudServerKey

**Review by KateVlaar — APPROVED** [review]

**Review by naviza — APPROVED** [review]
I'm excited!

**Pull request merged**